mdata/aggmetrics.go

```diff
@@ -87,18 +87,32 @@ func (ms *AggMetrics) Get(key schema.MKey) (Metric, bool) {
 }
 
 func (ms *AggMetrics) GetOrCreate(key schema.MKey, schemaId, aggId uint16) Metric {
-	ms.Lock()
+
+	// in the most common case, it's already there and an Rlock is all we need
+	ms.RLock()
 	m, ok := ms.Metrics[key]
-	if !ok {
-		k := schema.AMKey{
-			MKey: key,
-		}
-		agg := Aggregations.Get(aggId)
-		schema := Schemas.Get(schemaId)
-		m = NewAggMetric(ms.store, ms.cachePusher, k, schema.Retentions, schema.ReorderWindow, &agg, ms.dropFirstChunk)
-		ms.Metrics[key] = m
-		metricsActive.Set(len(ms.Metrics))
+	ms.RUnlock()
+	if ok {
+		return m
+	}
+
+	// if it wasn't there, get the write lock and prepare to add it
+	// but first we need to check again if someone has added it in
+	// the meantime
+	ms.Lock()
+	m, ok = ms.Metrics[key]
+	if ok {
+		ms.Unlock()
+		return m
+	}
+	k := schema.AMKey{
+		MKey: key,
 	}
+	agg := Aggregations.Get(aggId)
+	schema := Schemas.Get(schemaId)
```

Comment:
`Aggregations.Get()` and `Schemas.Get()` could both happen a little further up, after we know that we need to create an entry but before we acquire the write lock